File: CS50_2020/Week2/Readability/readability.c
#include<stdio.h>
#include<cs50.h>
#include<string.h>
#include<ctype.h>
#include<math.h>

// declare functions
int letterCount(string inputStr, int n);
int wordCount(string inputStr, int n);
int sentenceCount(string inputStr, int n);

// Main function here
int main(void)
{
    // get string from user
    string inputStr = get_string("Text: ");

    // number of chars in the string
    int n = strlen(inputStr);

    // number of letter per 100 words
    float L = letterCount(inputStr, n) * 100.0 / wordCount(inputStr, n);

    // number of sentences per 100 words
    float S = sentenceCount(inputStr, n) * 100.0 / wordCount(inputStr, n);

    // Coleman-Liau formula - readability index
    int index = round(0.0588 * L - 0.296 * S - 15.8);

    // check grade
    if (index <= 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %d\n", index);
    }
}

// Function for counting number of letters
int letterCount(string inputStr, int n)
{
    int letterCount = 0;
    for (int i = 0; i < n; i++)
    {
        // if ((inputStr[i] >=65 && inputStr[i] <= 90) | (inputStr[i] >=97 && inputStr[i] <= 122))
        if (isalpha(inputStr[i]))
        {
            letterCount++;
        }
    }
    return letterCount;
}

// Function for counting number of words
int wordCount(string inputStr, int n)
{
    int wordCount = 0;
    for (int i = 0; i < n; i++)
    {
        if (inputStr[i] == ' ')
        {
            wordCount++;
        }
    }
    // as one for the last word without space
    return wordCount + 1;
}

// Function for counting number of sentences
int sentenceCount(string inputStr, int n)
{
    int sentenceCount = 0;
    for (int i = 0; i < n; i++)
    {
        if (inputStr[i] == '.' | inputStr[i] == '!' | inputStr[i] == '?')
        {
            sentenceCount++;
        }
    }
    return sentenceCount;
}

File: CS50_2020/Week6/Readability/readability.py
from cs50 import get_string

string = get_string("Text: ")

length = len(string)

# initiallize letter, word and sentence counters
n_letter = 0
# words are seperated by a space, but the last word ends with . or ? or !
# therefore add 1 for initiallizing word count
n_words = 1
n_sentence = 0

# go through every character in the input text
for i in range(length):
    # check if the char is alphabetic
    if string[i].isalpha():
        n_letter += 1

    # words are seperated by a space
    if string[i] is ' ':
        n_words += 1

    # sentences are seperated by . ! or ?
    if string[i] in ['.', '!', '?']:
        n_sentence += 1

# number of letter and sentence per 100 words
L = n_letter * 100 / n_words
S = n_sentence * 100 / n_words

# Coleman-Liau formula - readability index
index = round(0.0588 * L - 0.296 * S - 15.8)

if index <= 1:
    print("Before Grade 1")
elif index >= 16:
    print("Grade 16+")
else:
    print("Grade " + str(index))